lang/java/reef-runtime-standalone/src/main/java/org/apache/reef/runtime/standalone/process/SshProcess.java

```diff
@@ -0,0 +1,330 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.runtime.standalone.process;
+
+import com.jcraft.jsch.*;
+import org.apache.reef.runtime.common.evaluator.PIDStoreStartHandler;
+import org.apache.reef.util.OSUtils;
+
+import java.io.*;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+/**
+ * A runnable class that encapsulates a process.
+ */
+public final class SshProcess implements Runnable {
+
+  private static final Logger LOG = Logger.getLogger(SshProcess.class.getName());
+
+  private static final long DESTROY_WAIT_TIME = 100;
+
+  /**
+   * Name of the file used for STDERR redirection.
+   */
+  private final String standardErrorFileName;
+  /**
+   * Name of the file used for STDOUT redirection.
+   */
+  private final String standardOutFileName;
+
+  /**
+   * Command to execute.
+   */
+  private final List<String> command;
+  /**
+   * User supplied ID of this process.
+   */
+  private final String id;
+  /**
+   * The working folder in which the process runs. It is also where STDERR and STDOUT files will be deposited.
+   */
+  private final File folder;
+  /**
+   * The coarse-grained lock for state transition.
+   */
+  private final Lock stateLock = new ReentrantLock();
+  private final Condition doneCond = stateLock.newCondition();
+  /**
+   * The process.
+   */
+  private Process process;
+  /**
+   * The state of the process.
+   */
+  private State state = State.INIT;   // synchronized on stateLock
+
+  private Session remoteSession;
+
+  private String remoteHostName;
+
+  private final String nodeFolder;
+
+  /**
+   * @param command               the command to execute.
+   * @param id                    The ID of the process. This is used to name files and in the logs created
+   *                              by this process.
+   * @param folder                The folder in which this will store its stdout and stderr output
+   * @param standardOutFileName   The name of the file used for redirecting STDOUT
+   * @param standardErrorFileName The name of the file used for redirecting STDERR
+   */
+  public SshProcess(final List<String> command,
```

Comment:
My understanding is that upon resource launch, you spawn a thread in `ProcessContainer` that runs `SshProcess`(a runnable), which in turn spawns a process that finally creates a ssh connection. Is this correct?

I think we can short-cut this by having each thread manage a ssh connection. This way, connection management and error handling would be much easier.